Standardize and add SSL settings

edmocosta opened the pull request:
### What this PR does?

Added the following SSL settings:
`ssl_enabled`: Enable/disable the SSL settings. If not provided, the value is inferred from the hosts scheme
`ssl_certificate`: OpenSSL-style X.509 certificate file to authenticate the client
`ssl_key`: OpenSSL-style RSA private key that corresponds to the `ssl_certificate`
`ssl_truststore_path`: he JKS truststore to validate the server's certificate
`ssl_truststore_type`: The format of the truststore file
`ssl_truststore_password`: The truststore password
`ssl_keystore_path`: The keystore used to present a certificate to the server
`ssl_keystore_type`: The format of the keystore file
`ssl_keystore_password`: The keystore password
`ssl_cipher_suites`: The list of cipher suites to use
`ssl_supported_protocols`: Supported protocols with versions
`ssl_verification_mode`: Defines how to verify the certificates presented by another party in the TLS connection

Reviewed and deprecated SSL settings to comply with Logstash's naming convention
`ssl` in favor of `ssl_enabled`
`ca_file` in favor of `ssl_certificate_authorities`
`keystore` in favor of `ssl_keystore_path`
`keystore_password` in favor of `ssl_keystore_password`

---
Closes https://github.com/elastic/logstash/issues/14923
Closes https://github.com/logstash-plugins/logstash-filter-elasticsearch/issues/164
Closes https://github.com/logstash-plugins/logstash-filter-elasticsearch/issues/49

andsel reviewed (changes requested):
Left some comments, that are quite similar to the observations done for output and input.

edmocosta commented on lib/logstash/filters/elasticsearch.rb:
```diff
@@ -219,16 +262,74 @@ def client_options
+    ssl_options[:enabled] = @ssl_enabled unless @ssl_enabled.nil?
+    ssl_options[:trust_strategy] = trust_strategy_for_ca_trusted_fingerprint
+
+    return ssl_options unless @ssl_enabled
```
Good catch! Setting the `trust_strategy` independently of the `ssl => true` option is actually a condition of the input plugin ([L439](https://github.com/logstash-plugins/logstash-input-elasticsearch/blob/main/lib/logstash/inputs/elasticsearch.rb#L439)). I've mistakenly applied it here as well. 

It was also missing an extra condition to keep it backward compatible. I've pushed the changes and added a comment to explain the intention. The main idea is to have the same functionality for the `ssl_enabled` flag across the three Elasticsearch plugins.

andsel commented on lib/logstash/filters/elasticsearch.rb:
```diff
@@ -219,16 +264,79 @@ def client_options
+    ssl_certificate_authorities, ssl_truststore_path, ssl_certificate, ssl_keystore_path = params.values_at('ssl_certificate_authorities', 'ssl_truststore_path', 'ssl_certificate', 'ssl_keystore_path')
+
+    if ssl_certificate_authorities && ssl_truststore_path
+      raise(LogStash::ConfigurationError, 'Use either "ssl_certificate_authorities/ca_file" or "ssl_truststore_path" when configuring the CA certificate')
```
```suggestion
      raise LogStash::ConfigurationError, 'Use either "ssl_certificate_authorities/ca_file" or "ssl_truststore_path" when configuring the CA certificate'
```

andsel commented on lib/logstash/filters/elasticsearch.rb:
```diff
@@ -219,16 +264,79 @@ def client_options
+    end
+
+    if ssl_certificate && ssl_keystore_path
+      raise(LogStash::ConfigurationError, 'Use either "ssl_certificate" or "ssl_keystore_path/keystore" when configuring client certificates')
```
```suggestion
      raise LogStash::ConfigurationError, 'Use either "ssl_certificate" or "ssl_keystore_path/keystore" when configuring client certificates'
```

andsel commented on lib/logstash/filters/elasticsearch.rb:
```diff
@@ -219,16 +264,79 @@ def client_options
+    end
+
+    if ssl_certificate_authorities&.any?
+      raise(LogStash::ConfigurationError, 'Multiple values on "ssl_certificate_authorities" are not supported by this plugin') if ssl_certificate_authorities.size > 1
```
```suggestion
      raise LogStash::ConfigurationError, 'Multiple values on "ssl_certificate_authorities" are not supported by this plugin' if ssl_certificate_authorities.size > 1
```

andsel commented on lib/logstash/filters/elasticsearch.rb:
```diff
@@ -219,16 +264,79 @@ def client_options
+      ssl_options[:client_cert] = ssl_certificate
+      ssl_options[:client_key] = ssl_key
+    elsif !!ssl_certificate || !!ssl_key
+      raise(LogStash::ConfigurationError, 'You must set both "ssl_certificate" and "ssl_key" for client authentication')
```
```suggestion
      raise LogStash::ConfigurationError, 'You must set both "ssl_certificate" and "ssl_key" for client authentication'
```

andsel reviewed (changes requested):
Left a nitpick style note

edmocosta commented:
> Left a nitpick style note

Thank you again for reviewing it, @andsel. I've addressed your suggestions plus the [suggestion #1](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1118#discussion_r1131890635 ) and [suggestion #2](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1118#discussion_r1131892391) made on the output plugin.

andsel reviewed (approved):
LGTM

Just update the 
```
deprecated[8.8.0, Replaced by
```
 to refer to the plugin's version instead of Logstash one.

kaisecheng commented on CHANGELOG.md:
```diff
@@ -1,3 +1,23 @@
+    - `ssl_enabled`: Enable/disable the SSL settings. If not provided, the value is inferred from the hosts scheme
+    - `ssl_certificate`: OpenSSL-style X.509 certificate file to authenticate the client
+    - `ssl_key`: OpenSSL-style RSA private key that corresponds to the `ssl_certificate`
+    - `ssl_truststore_path`: he JKS truststore to validate the server's certificate
```
```suggestion
    - `ssl_truststore_path`: The JKS truststore to validate the server's certificate
```

edmocosta commented on lib/logstash/filters/elasticsearch.rb:
```diff
@@ -219,16 +264,80 @@ def client_options
+
+    ssl_key = params["ssl_key"]
+    if ssl_certificate
+      raise LogStash::ConfigurationError, 'Using an "ssl_certificate" requires an "ssl_key"' unless ssl_key
```
I'm gonna move on with this PR and merge it! If that's a mistake I can submit another PR to fix it! Thanks!

djschny commented:
How can I tell what version of Elastic stack this is in?

tommycahir commented:
I am just running v3.15.2 and I am seeing an error saying " Unknown setting 'ssl_certificate_verification' for elasticsearch" when using this filter

edmocosta commented:
Hi @tommycahir!

The `ssl_certificate_verification` setting does not exist for this filter. I think the correct one is [ssl_verification_mode](https://www.elastic.co/guide/en/logstash/current/plugins-filters-elasticsearch.html#plugins-filters-elasticsearch-ssl_verification_mode) :)
